### PR TITLE
Delete repos of org when purge delete user

### PR DESCRIPTION
**Pull request by JakobDev**
Fixes https://codeberg.org/forgejo/forgejo/issues/1514

I had to remove `RenameOrganization` to avoid circular import.

We should really add some foreign keys to the database.

### CODE REVIEW

File: services/org/org.go [hunk 1]
```diff
@@ -15,12 +15,44 @@ import (
...
+			}
+			for _, repo := range repos {
+				if err := repo_service.DeleteRepositoryDirectly(ctx, org.AsUser(), org.ID, repo.ID); err != nil {
+					return fmt.Errorf("unable to delete repository %s for %s[%d]. Error: %w", repo.Name, org.Name, org.ID, err)
+				}
+			}
+		}
+	}
+
+	ctx, commiter, err := db.TxContext(ctx)
```
Comment (lunny) [inline services/org/org.go:23]: I think `DeleteRepositoryDirectly` should be also under the transaction. Otherwise, if there will be a dirty situation that all repositories deleted but organization related information hasn't been deleted.

**Review by lunny — DISMISSED** [review]
https://github.com/go-gitea/gitea/pull/27273/files#r1337160163

File: services/org/org.go (hunk 1 above)
Comment (JakobDev) [reply services/org/org.go:23]: `DeleteRepositoryDirectly` uses a transaction, so we would have a transaction in a transaction. Not sure if that's causing problems.

Comment (wojtekLs) [reply services/org/org.go:55]: @JakobDev nested transactions in `sql` are normal thing.

File: services/user/user.go [hunk 2]
```diff
@@ -206,7 +207,10 @@ func DeleteUser(ctx context.Context, u *user_model.User, purge bool) error {
 			for _, org := range orgs {
 				if err := models.RemoveOrgUser(org.ID, u.ID); err != nil {
 					if organization.IsErrLastOrgOwner(err) {
-						err = organization.DeleteOrganization(ctx, org)
+						err = org_service.DeleteOrganization(ctx, org, true)
+						if err != nil {
+							return fmt.Errorf("unable to delete organisation %d: %w", org.ID, err)
```
Comment (delvh) [inline services/user/user.go]:
```suggestion
							return fmt.Errorf("unable to delete organization %d: %w", org.ID, err)
```

File: services/org/org.go [hunk 3]
```diff
@@ -15,12 +15,44 @@ import (
...
+			}
+			if len(repos) == 0 {
+				break
+			}
+			for _, repo := range repos {
+				if err := repo_service.DeleteRepositoryDirectly(ctx, org.AsUser(), org.ID, repo.ID); err != nil {
+					return fmt.Errorf("unable to delete repository %s for %s[%d]. Error: %w", repo.Name, org.Name, org.ID, err)
+				}
+			}
+		}
```
Comment (delvh) [inline services/org/org.go]: I recommend a separate function instead of the nesting.

File: services/org/org.go (hunk 1 above)
Comment (JakobDev) [reply services/org/org.go:23]: I have now tried to call `DeleteRepositoryDirectly` in the transaction, but it resulted in a deadlock in SQLite

Comment (wojtekLs) [reply services/org/org.go:30]: @JakobDev anything out of ordinary in log?

Comment (JakobDev) [reply services/org/org.go:23]: No, it just hangs forever. If you are interested, you can easily try it out yourself.

Comment (lunny) [reply services/org/org.go:23]: > No, it just hangs forever. If you are interested, you can easily try it out yourself.

`ctx, commiter, err := db.TxContext(ctx)` will check whether `ctx` contains a transaction, if it contains, it will reuse the transaction, otherwise it will create one. This has been used many places in Gitea codebase.

Comment (JakobDev) [reply services/org/org.go:23]: It is used in may places, but this particular place causes a deadlock, so we can't use it here.

File: services/repository/delete.go [hunk 4]
```diff
@@ -422,3 +422,30 @@ func RemoveRepositoryFromTeam(ctx context.Context, t *organization.Team, repoID
 
 	return committer.Commit()
 }
+
+// DeleteOwnerRepositoriesDirectly calls DeleteRepositoryDirectly for all repos of the given owner
+func DeleteOwnerRepositoriesDirectly(ctx context.Context, owner *user_model.User) error {
+	for {
+		repos, _, err := repo_model.GetUserRepositories(&repo_model.SearchRepoOptions{
```
Comment (lunny) [inline services/repository/delete.go]: This function needs a ctx, I think it's the reason why deadlock.

Comment (JakobDev) [reply services/repository/delete.go]: You're right. That solved the problem.

File: services/org/org.go [hunk 5]
```diff
@@ -15,12 +15,19 @@ import (
...
-	user_service "code.gitea.io/gitea/services/user"
+	repo_service "code.gitea.io/gitea/services/repository"
 )
 
 // DeleteOrganization completely and permanently deletes everything of organization.
-func DeleteOrganization(org *org_model.Organization) error {
-	ctx, commiter, err := db.TxContext(db.DefaultContext)
+func DeleteOrganization(ctx context.Context, org *org_model.Organization, purge bool) error {
+	if purge {
+		err := repo_service.DeleteOwnerRepositoriesDirectly(ctx, org.AsUser())
```
Comment (delvh) [inline services/org/org.go:30]:
```suggestion
		// TODO: Put this into the transaction below, for some reason it currently causes a deadlock (SQLite)
		err := repo_service.DeleteOwnerRepositoriesDirectly(ctx, org.AsUser())
```

**Review by delvh — APPROVED** [review]

File: services/org/org.go (hunk 5 above)
Comment (JakobDev) [reply services/org/org.go:30]: The deadlock problem was already resolved

**Comment by JakobDev** [comment]
Bump

**Comment by 6543** [comment]
could you resolve the conflicts please?

File: services/org/org.go [hunk 6]
```diff
@@ -67,8 +74,3 @@ func DeleteOrganization(ctx context.Context, org *org_model.Organization) error
 
 	return nil
 }
-
-// RenameOrganization renames an organization.
```
Comment (6543) [inline services/org/org.go:71]: unrelated refactor !!!

File: routers/web/org/setting.go [hunk 7]
```diff
@@ -75,7 +75,7 @@ func SettingsPost(ctx *context.Context) {
 
 	// Check if organization name has been changed.
 	if nameChanged {
-		err := org_service.RenameOrganization(ctx, org, form.Name)
```
Comment (6543) [inline routers/web/org/setting.go:78]: unrelated refactor

**Review by 6543 — CHANGES_REQUESTED** [review]

**Comment by 6543** [comment]
haha that dependency cycle got me 😅

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
hmm ci fail looks related

```
2023/10/16 20:13:21 ...es/repository/lfs.go:74:GarbageCollectLFSMetaObjectsForRepo() [E] Unable to open git repository <Repository 1:user2/repo1>: no such file or directory
--- FAIL: TestGarbageCollectLFSMetaObjects (0.03s)
    lfs_test.go:43: 
        	Error Trace:	/home/runner/work/gitea/gitea/services/repository/lfs_test.go:43
        	Error:      	Received unexpected error:
        	            	no such file or directory
        	Test:       	TestGarbageCollectLFSMetaObjects
    lfs_test.go:47: 
        	Error Trace:	/home/runner/work/gitea/gitea/services/repository/lfs_test.go:47
        	Error:      	Target error should be in err chain:
        	            	expected: "LFS Meta object does not exist"
        	            	in chain: 
        	Test:       	TestGarbageCollectLFSMetaObjects
FAIL
coverage: 22.8% of statements
```

@JakobDev what do you think

**Comment by JakobDev** [comment]
The test is now fixed. It was caused by a Bug in `TestGarbageCollectLFSMetaObjects`. This tests needs the Repos, but don't call a function to make sure the Repos exists.

**Comment by JakobDev** [comment]
This should be backported to 1.21

**Comment by 6543** [comment]
->  #27728
